### ISSUE TITLE
OData v4 datasource transport serializes date filters without converting to UTC

### ISSUE THREAD

**Issue by ggkrustev** [telerik/kendo-ui-core#1864]
### Description

The built-in 'odata-v4' transport type serializes the filter date values using a ISO8601 format with dedicated timezone offset (+0000 in this case). It seems, however, that the date is not converted to UTC timezone before formatting it, hence the timezone offset of the date wind up to be incorrect.

The current code does not adjust the timezone offset, before formatting it.

```
format = "{1:yyyy-MM-ddTHH:mm:ss+00:00}";
```
### Solution

We will need to adjust the date before passing it to the format. The code that should be changed is:
- https://github.com/telerik/kendo-ui-core/blob/master/src/kendo.data.odata.js#L149

Basically, we would need to call `kendo.timezone.apply(value, "Etc/UTC")` and then to pass the `value`.
### Backward compatibility

The fix will change current ODATA v4 filter serialization behavior. That is why we will include the fix in Q3 2016 release.

**Comment by PoulTur** [comment]
I see in [the 2016 Q3 release history](http://www.telerik.com/support/whats-new/kendo-ui/release-history/kendo-ui-r3-2016) and likely linked to this thread here about fixes in dataSource linked to dates. Does this mean that kendo dataSource now supports (fully) OData v4. unlike in [docs] 
(http://docs.telerik.com/kendo-ui/api/javascript/data/datasource#configuration-type) ?

**Comment by pekuz** [comment]
I face tz problems in Kendo UI v2016.1.112. In fact date filters already have UTC/Z Date values and then translation to odata-v4 shifts the instant towards UTC/Z once again.

Example: being in UTC+1, entering 2016-11-03 in Kendo grid date filter, there is "2016-11-02T23:00:00.000Z" Date before passing to odata-04 transport.parameterMap() that translates to Odata $filter that contains unexpected "2016-11-03T00:00:00+00:00".  Now if I before odata-04 transport.parameterMap() shift by 60 minutes i.e. having "2016-11-02T22:00:00.000Z" Date before passing to odata-04 transport.parameterMap() then it translates to Odata $filter that contains expected "2016-11-02T23:00:00+00:00".

So in my experience default Kendo odata-04 transport.parameterMap() impl does time shifting but it shifts  by UTC/Z offset even values that are already in UTC/Z tz. What is design intent there, please? What should work in end user tz and what piece of code is responsible for converting to UTC/Z for Odata-v4 transport, please?

**Comment by Alex-Bubblemaster** [comment]
Ticket ID: 1404615 

Currently the `odata-v4` transport type offsets the dates to UTC according to the timezone of the client:
![image](https://user-images.githubusercontent.com/10980842/56221688-61cf6c80-6073-11e9-827a-3c7f10ab7bcb.png)

But instead, it should be sending a new date in UTC without the offset - with 00:00:00 time, like this:
![image](https://user-images.githubusercontent.com/10980842/56221790-90e5de00-6073-11e9-9583-ef4f57f6b8ce.png)

While we do not have a service, here is an example with server filtering, you can filter the date field and check the request that is sent. [Dojo to reproduce](https://dojo.telerik.com/@bubblemaster/UBicekOW)

### Workaround

Use the `transport.read.data` to create a UTC date, e.g:

[Dojo](https://dojo.telerik.com/@bubblemaster/UBicekOW)

```
data: function(e){
  if(e.filter){
    var filters = e.filter.filters;
    for(var i=0; i < filters.length; i++){
      if(filters[i].field === "OrderDate"){
        var selectedDate =  filters[i].value;
       e.filter.filters[i].value = new Date(Date.UTC(selectedDate.getFullYear(), selectedDate.getMonth(), selectedDate.getDate()));
      }
    }
  }
}
```

**Comment by IvanDanchev** [comment]
Priority raised (ID: 1463411)

**Comment by gdenchevprog** [comment]
Raising priority (ID: 1528561)

**Comment by gdenchevprog** [comment]
The behavior that occurs after the initial fix is the correct one, here's the explanation:

The date has always been formatted with "+00:00" at the end. This indicates that the date is in UTC. Before the fix, a user could send a date from their own timezone (let's say +02:00), however the final format would still have "+00:00" at the end. This means that the date is stored as if it's in UTC when it is not.

The fix addresses this issue by always applying the UTC time zone to the date that comes from the user(which is in their local time).

This means that if the user is in a timezone that is UTC+2 and they select a date 2023-31-01, the selected date will be based on their local time. The actual UTC date should be UTC+0 which means 2 hours must be subtracted in order to get the correct result.

Saving the date using UTC+0 is desirable as you may have users from different timezones. If this causes problems with the reverse process(getting the date from the database and showing it to the user) then that is where the problem should be addressed.

The displayed date should be formatted using the users local timezone(if that is the desired outcome).

If a different type of formatting is required in a specific situation, then you can use the [parameterMap](https://docs.telerik.com/kendo-ui/api/javascript/data/datasource/configuration/transport.parametermap) configuration to modify the request parameters.
